ramlfications/types.py

```diff
@@ -0,0 +1,192 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2015 Ramlfications contributors
+
+from __future__ import absolute_import, division, print_function
+import attr
+import copy
+from six import MAXSIZE
+from six import iteritems
+from six import string_types
+
+from ramlfications.errors import UnknownDataTypeError
+from ramlfications.errors import DataTypeValidationError
+from ramlfications.parameters import Content
+from ramlfications.utils.common import OrderedDict
+from ramlfications.utils.parser import convert_caml_case
+import re
+
+__type_registry = {}
+
+
+def type_class(name):
+    def decorator(klass):
+        __type_registry[name] = klass
+        return klass
+    return decorator
+
+
+# In order to simplify the code, we match 1:1 the raml type definition to the
+# attrs object representation
+def create_type(name, raml_def):
+    # spec:
+    # string is default type when nothing else defined
+    typeexpr = raml_def.get('type', 'string')
+
+    if typeexpr not in __type_registry:
+        # we start simple, type expressions are for another commit
+        raise UnknownDataTypeError(
+            "{0} type expression is not supported or not defined".format(
+                typeexpr))
+
+    klass = __type_registry[typeexpr]
+    print(klass)
+    # we try here to be very near the raml to avoid lots of boilerplate code to
+    # convert raml spec into our internal object format
+    # remaining conversion is handle via the 'convert=' parameters of the attrs
+    # library
+
+    raml_def = dict([
+        (convert_caml_case(k), v) for k, v in iteritems(raml_def)
+    ])
+    return klass(name=name, **raml_def)
+
+
+@attr.s
+class BaseType(object):
+    """
+    Base class for all raml data types.
+
+    :param string name: name of the type
+    :param Content description: description for the type.
+        This is a markdown Content with on the fly conversion to html using
+        description.html
+    :type string base type for this type
+    """
+    name            = attr.ib()
+    description     = attr.ib(default="", repr=False, convert=Content)
+    type            = attr.ib(default="string", repr=False)
+
+    def validate(self, validated_objet, position_hint=None):
```

Comment:
Having these validate methods goes a bit against the already defined validation pattern. What is the purpose of having these attached to the object, rather than in `ramlfications/validate.py`?